### ISSUE TITLE
[Bug]: Inconsistent version when install 3.5.0 to 3.5.2 on GitHub CI

### ISSUE THREAD

**Issue by oscargus** [matplotlib/matplotlib#23325]
### Bug summary

Not sure if this is an MPL issue, but here goes.

The SymPy tests cannot install the latest pypi versions of MPL as the message `filename has '3.5.2', but metadata has '0.0'` is obtained. See more below. (It ends up installing 3.4.3.)

### Code for reproduction

```python
pip install matplotlib
```

### Actual outcome

```
Collecting matplotlib
  Downloading matplotlib-3.5.2.tar.gz (35.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.2/35.2 MB 56.8 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Discarding https://files.pythonhosted.org/packages/2f/be/7d6e073a3eb740ebeba43a69f5de2b141fea50b801e24e0ae024ac94d4ac/matplotlib-3.5.2.tar.gz#sha256=48cf850ce14fa18067f2d9e0d646763681948487a8080ec0af2686468b4607a2 (from https://pypi.org/simple/matplotlib/) (requires-python:>=3.7): Requested matplotlib from https://files.pythonhosted.org/packages/2f/be/7d6e073a3eb740ebeba43a69f5de2b141fea50b801e24e0ae024ac94d4ac/matplotlib-3.5.2.tar.gz#sha256=48cf850ce14fa18067f2d9e0d646763681948487a8080ec0af2686468b4607a2 has inconsistent version: filename has '3.5.2', but metadata has '0.0'
```

### Expected outcome

metadata is not 0.0

### Matplotlib Version

3.5.0 to 3.5.2

**Comment by tacaswell** [comment]
This is likely something wrong with our build process.  Just last night I noticed something wonking with setuptools_scm 7.0.1, but have not been able to run that to ground at all.

**Comment by oscarbenjamin** [comment]
To be clear this problem is only seen on SymPy's Python 3.11 bleeding edge CI job which uses cython, numpy and scipy from git master and then tries to install various other things including matplotlib with pip. You can see this setup here:
https://github.com/sympy/sympy/blob/e31373d33ddd4d80a77ebc68e29b7fafb59055a6/.github/workflows/runtests.yml#L167-L209

In https://github.com/sympy/sympy/pull/23684 I am changing it to not install matplotlib there. We could alternatively switch to using matplotlib master.

There are no 3.11 wheels given on PyPI for matplotlib 3.5.2 so I guess it's the sdist that leads to this problem.

I tried under Python 3.8 to install the sdist and I see this:
```python
$ pip install ~/Downloads/matplotlib-3.5.2.tar.gz 
Processing /home/oscar/Downloads/matplotlib-3.5.2.tar.gz
Requirement already satisfied: cycler>=0.10 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (0.10.0)
Collecting fonttools>=4.22.0
  Using cached fonttools-4.33.3-py3-none-any.whl (930 kB)
Requirement already satisfied: kiwisolver>=1.0.1 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (1.2.0)
Requirement already satisfied: numpy>=1.17 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (1.18.3)
Requirement already satisfied: packaging>=20.0 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (20.9)
Requirement already satisfied: pillow>=6.2.0 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (8.2.0)
Requirement already satisfied: pyparsing>=2.2.1 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (2.4.7)
Requirement already satisfied: python-dateutil>=2.7 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (2.8.1)
Requirement already satisfied: six in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from cycler>=0.10->matplotlib==0.0) (1.14.0)
Using legacy 'setup.py install' for matplotlib, since package 'wheel' is not installed.
Installing collected packages: fonttools, matplotlib
    Running setup.py install for matplotlib ... ... done
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
arviz 0.11.2 requires matplotlib>=3.0, but you have matplotlib 0.0 which is incompatible.
Successfully installed fonttools-4.33.3 matplotlib-0.0
```
So you can see that pip thinks that the version for this sdist is 0.0. That was without the wheel package installed which uses the `setup.py` method of installation. With wheel installed I get the same though.

Inside the sdist in the setup.py I can see this:
```python
    use_scm_version={
        "version_scheme": "release-branch-semver",
        "local_scheme": "node-and-date",
        "write_to": "lib/matplotlib/_version.py",
        "parentdir_prefix_version": "matplotlib-",
        "fallback_version": "0.0+UNKNOWN",
    },
```
I can see a `0.0` there which perhaps explains where that comes from. I have no idea how `use_scm_version` is supposed to work though. The `_version.py` file seems to be correct but I guess it is being rewritten somehow during build::
```console
$ cat lib/matplotlib/_version.py 
# coding: utf-8
# file generated by setuptools_scm
# don't change, don't track in version control
version = '3.5.2'
version_tuple = (3, 5, 2)
```
After install I have:
```python
In [1]: import matplotlib                                                                            

In [2]: matplotlib.__version__                                                                       
Out[2]: '0.0'
```

**Comment by oscarbenjamin** [comment]
Installing latest master from a git checkout seems to work fine with various invocations like
```console
$ python setup.py install
$ pip install .
```
Those all give a version like `3.6.0.dev2573+g3eadeac`. However doing the same with the 3.5.2 sdist always seems to give version 0.0.

I tried creating the sdist from git:
```console
$ python setup.py sdist
```
Then installing from there gives the same problem:
```console
$ pip install matplotlib/dist/matplotlib-3.6.0.dev2573+g3eadeac.tar.gz 
Processing ./matplotlib/dist/matplotlib-3.6.0.dev2573+g3eadeac.tar.gz
  Preparing metadata (setup.py) ... done
Requirement already satisfied: contourpy>=1.0.1 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages/contourpy-1.0.3-py3.8-linux-x86_64.egg (from matplotlib==0.0) (1.0.3)
Requirement already satisfied: cycler>=0.10 in /home/oscar/current/sympy/38venv/lib/python3.8/site-packages (from matplotlib==0.0) (0.10.0)
...
```

**Comment by oscargus** [comment]
> So probably this is the same issue as [SciTools/cartopy#2053](https://github.com/SciTools/cartopy/issues/2053)?

It does indeed look like the same type of error.

**Comment by tacaswell** [comment]
That also explains why pushing setuptools_scm to<7 "fixed" it for me locally.

**Comment by dstansby** [comment]
Looks like this might be https://github.com/pypa/setuptools_scm/issues/745?

**Comment by QuLogic** [comment]
Should actually fixed in 3.5.3 by #23479.
